### PR TITLE
Add option for the Mu modes to attempt to open a file

### PR DESCRIPTION
**Pull request by microbit-carlos**
Related to https://github.com/mu-editor/mu/pull/452, I wasn't sure if this approach was preferable instead of just doing the check in `Editor._load()`, so I thought I should check before I update the tests.

In essence, this extends the `BaseMode` class to have an `open` method, so `Editor._load()` only tries to open `.py` files, if the selected file has a different extension it then goes through each mode and sees if they can open the file.

It a mode opens the file it then asks the user if they'd like to change to that mode (if they are not there already).

### CODE REVIEW

**Comment by ZanderBrown** [comment]
Cool idea, perhaps we could have custom editor panes as well? Perhaps that should wait for 1.1.....

File: mu/interface/main.py [hunk 1]
```diff
@@ -224,13 +224,13 @@ def set_read_only(self, is_readonly):
         for tab in self.widgets:
             tab.setReadOnly(is_readonly)
 
-    def get_load_path(self, folder):
+    def get_load_path(self, folder, extensions):
```
Comment (carlosperate) [inline mu/interface/main.py]: This one should have the default value: `extensions='*'`

File: mu/modes/base.py [hunk 2]
```diff
@@ -176,6 +177,12 @@ def on_data_flood(self):
                  "time.")
         self.view.show_message(msg, info)
 
+    def open_file(self, path):
+        """
+        Some files are not plain text and each mode can attempt to decode them.
+        """
+        return None
```
[the remark this answers is not in the document]
Comment (microbit-carlos) [reply mu/modes/base.py]: Actually my reasoning was that if the method returns None then it means it cannot open the input file path.
This way we can count on all modes containing this method (and not having to check if is implemented before calling it) and that it returns a "valid" value (which can be None), even if the subclasses don't implement it.
Should we instead throw a `NotImplemented` and force all modes to have it?

[the remark this answers is not in the document]
Comment (ntoll) [reply mu/modes/base.py]: No, having looked at the code in the wider context I think this is probably the right thing to do. :-)

**Comment by microbit-carlos** [comment]
This should be ready once the CI tests have passed 🤞

**Comment by carlosperate** [comment]
@ntoll looks like this branch now has conflicts with master, should I rebase or wait a bit more?

**Comment by ntoll** [comment]
Please re-base. My changes surrounded the fact that Adafruit mode had lots of pop-ups happening and would also crash in certain conditions, when re-starting Mu with the device NOT plugged in.

**Comment by carlosperate** [comment]
@ntoll, sorry forgot to comment on this on Tuesday to say it was rebased.

**Comment by ntoll** [comment]
@carlosperate first of all, apologies for the tardy response time on my part (blame, PyCon, training in Ireland and family commitments for my lack of time over the past month). Secondly, this is great, awesome work..! Thank you! Finally, I'm thinking of adding a "Documentation" mode post 1.0 (i.e. so Mu understands Markdown) and this is a great first step in this direction. :-)

**Comment by microbit-carlos** [comment]
No worries, thanks Nicholas!

**Pull request merged**